apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,230 @@
+'use strict';
+
+/**
+ * Class which allows callers to listen to changes to data tables. This is currently
+ * done by polling. Some modest rate-limiting is achieved by waiting for one second
+ * after the previous response is processed for data from a given table, rather than
+ * polling on a fixed interval and allowing a backlog to accumulate.
+ *
+ * The most logical incremental performance improvement is to use pusher to only
+ * re-read from the table if any of its contents have changed. Down the road, we
+ * may be able to use DynamoDB Streams / Lambda / SQS to deliver fine-grained
+ * notifications to the client without ever needing to  re-read the entire table.
+ *
+ * @constructor
+ */
+var RecordListener = module.exports = function () {
+  /**
+   * Map from table name to table handler.
+   * @private {Object.<string, TableHandler>}
+   */
+  this.tableHandlers_ = {};
+};
+
+/**
+ * Number of ms to wait before polling each table for more data after finishing
+ * processing the last batch of data for that table.
+ * @type {number}
+ */
+var RECORD_INTERVAL = 1000; // 1 second
+
+var EventType = {
+  CREATE: 'create',
+  UPDATE: 'update',
+  DELETE: 'delete'
+};
+RecordListener.EventType = EventType;
+
+/**
+ * Adds a listener which calls the callback when changes are made to the table.
+ * The callback is called once per changed record.
+ * @param {string} tableName Name of the table to listen to.
+ * @param {function(Object, EventType)} callback Callback to call with the record
+ * and the type of event which happened to it. For create and update events,
+ * the entire new record is included. For delete events, only the id is included.
+ * @returns {boolean} true if adding the listener succeeded, or false if
+ * a listener already existed for the specified table.
+ */
+RecordListener.prototype.addListener = function (tableName, callback) {
+  if (this.tableHandlers_[tableName]) {
+    return false;
+  }
+
+  this.tableHandlers_[tableName] = new TableHandler(tableName, callback);
+
+  return true;
+};
+
+/**
+ * Clears the timeouts and resets internal state.
```

Comment:
Is this call intended to be part of the public API?  If it is, please add a description that is less implementation dependent;  the caller shouldn't have to know what you mean by "timeouts".